core/rawdb/freezer_batch.go

```diff
@@ -182,13 +188,108 @@ func (batch *freezerBatch) write(newHead bool) (bool, error) {
 		// Some data left to write on a retry.
 		batch.sizes = batch.sizes[count:]
+		batch.total = 0
+		for _, size := range batch.sizes {
+			batch.total += int(size)
+		}
 		return true, nil
 	}
 	// All data written. We can simply truncate and keep using the buffer
 	batch.sizes = batch.sizes[:0]
+	batch.total = 0
 	return false, nil
 }
 
-func (batch *freezerBatch) Size() int {
-	return batch.buf.Len()
+func (batch *freezerTableBatch) Size() int {
+	return batch.total
+}
+
+// freezerBatch is a write-only database that commits changes to the associated
+// ancient store when Write is called. A freezerBatch cannot be used concurrently.
+type freezerBatch struct {
+	lock    sync.RWMutex
+	f       *freezer
+	batches map[string]*freezerTableBatch // List of table batches
+	size    int
+}
+
+// AppendAncient injects all binary blobs belong to block into the batch.
+func (batch *freezerBatch) AppendAncient(number uint64, hash, header, body, receipt, td []byte) error {
+	batch.lock.RLock()
+	defer batch.lock.RUnlock()
+
+	if err := batch.batches[freezerHeaderTable].Append(number, header); err != nil {
+		return err
+	}
+	if err := batch.batches[freezerHashTable].Append(number, hash); err != nil {
+		return err
+	}
+	if err := batch.batches[freezerBodiesTable].Append(number, body); err != nil {
+		return err
+	}
+	if err := batch.batches[freezerReceiptTable].Append(number, receipt); err != nil {
+		return err
+	}
+	if err := batch.batches[freezerDifficultyTable].Append(number, td); err != nil {
+		return err
+	}
+	return nil
+}
+
+// TruncateAncients discards all but the first n ancient data from the ancient store.
+func (batch *freezerBatch) TruncateAncients(n uint64) error {
+	return errNotSupported
+}
+
+// Sync flushes all in-memory ancient store data to disk.
+func (batch *freezerBatch) Sync() error {
+	return errNotSupported
+}
+
+// ValueSize retrieves the amount of data queued up for writing.
+func (batch *freezerBatch) ValueSize() int {
+	batch.lock.RLock()
+	defer batch.lock.RUnlock()
+
+	var size int
+	for _, batch := range batch.batches {
+		size += batch.Size()
+	}
+	return size
+}
+
+// Write flushes any accumulated data to ancient store.
+func (batch *freezerBatch) Write() error {
+	if batch.f.readonly {
+		return errReadOnly
+	}
+	batch.lock.RLock()
+	defer batch.lock.RUnlock()
+
+	if err := batch.batches[freezerHeaderTable].Write(); err != nil {
+		return err
+	}
+	if err := batch.batches[freezerHashTable].Write(); err != nil {
+		return err
+	}
+	if err := batch.batches[freezerBodiesTable].Write(); err != nil {
+		return err
+	}
+	if err := batch.batches[freezerReceiptTable].Write(); err != nil {
+		return err
+	}
+	if err := batch.batches[freezerDifficultyTable].Write(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Reset resets the batch for reuse.
+func (batch *freezerBatch) Reset() {
+	batch.lock.Lock()
+	defer batch.lock.Unlock()
+
+	for name, table := range batch.f.tables {
+		batch.batches[name] = table.NewBatch()
```

Comment:
You should not need to do this. After writing the table-batches, they should be already reset and ready for reuse. So the locking isn't needed here